### DISCUSSION TITLE
Draw inputs while playing replays

### DISCUSSION THREAD

**Discussion by Mrcubix** [ppy/osu#26856] [Ideas] [answered]
All that is currently missing to make rewind obsolete is the ability to see the flow of the cursor as well as each and every taps done by each fingers while watching a replay.

![image](https://github.com/ppy/osu/assets/39861216/eb7a9cc9-c183-44bf-a630-8096e9bf6e9e)

The simple way of implementing it would be to just draw lines between each inputs and draw cross of different colors depending on which finger tapped on a key at a time t.

**Comment by Sheppsu** [comment]
I can work on this and post a video for feedback once I have it working. If it gets to a good state I'll make a PR.

**Reply by Sheppsu** [reply]
It's in an early stage right now, but feel free to give feedback. I already know that it needs some mod disabling functionality like rewind for mods that alter the hit circles (specifically hidden). The cursor hide option is also only half-working (I'm too tired to fix it now). The Xs stay on screen for 1 seconds each, fading in the last 0.4 seconds, which was just an arbitrary choice. Orange Xs are left clicks and green is right click. One last thing: the replay analysis settings box is mode specific, although I've only implemented one for standard. The fork/branch is [here](https://github.com/Sheppsu/osu/tree/replay-analysis-settings).

https://github.com/ppy/osu/assets/49356627/c3b8837d-e8f4-4124-a9f5-7951b8839da9

**Reply by Mrcubix** [reply]
The concept is definitely there.
You may want to provide user settings for fade out & the time they stay on screen, if that doesn't complicate things too much.
It does need some visual polish of course but that can be done is due time, great progress so far!

**Reply by yuutsutv** [reply]
Ok, so I have a few (a lot) of thoughts, suggestions, and opinions, which you don't need to follow up on in any way. I just personally think these things.

- Could you have the hit markers skinnable?
Maybe allow the player to implement a circle/dot for the marker instead of a X?

- Is there anyway to make the cursor pathing a solid line?
When using the mini X's for crossover patterns it could get pretty confusing pretty quick, maybe also make this skinnable?

- Could you make the hit markers match the color to the accuracy?
300's blue, 100's green, etc? It seems like (although I could be wrong) that the 300's are gold?

- Maybe a slider to adjust the size of the hitmarkers on the fly to show a more precise hit?

Issues I see:
when you pause and rewind in the video at 0:27, it seems the note moves? Maybe that's a replay file issue, idk.

I also have another separate idea to this but somewhat functionally the same.
At the end screen could there be a button down next to download/view replay and restart to show frames kinda like Miss Analyzer bot in discord does? It shows neat breakdowns of notes, timing of each individual miss, etc?

It's really nice to be able to see exactly how you missed and what the timing breakdown was on each individual note.
Maybe this could be implemented as an element in the replay analyzer itself?
Hope this helps!

**Reply by Mrcubix** [reply]
- Should not be a priority, after all it's not skinnable in rewind,
- ~~it's already a solid line i believe? Just many report leads to many aim marker,~~ nvm, there should be one in due time, to match rewind.
- The hit marker colors are here to indicate whether it's a key 1 tap or a key 2 tap,
- Not really needed either, it's already precise enough, but i guess that could be available.

**Reply by yuutsutv** [reply]
- Why not make it skinnable? Is there any reason not to? Circles would be a lot easier to see considering the aim markers are already X spam, makes it weird to read.
- There is a gap between the aim markers with no line, maybe in your version. But sheppsu’s version here doesn’t have it. My comment was based on the video in the above comment not on your original post.
- What’s the point in key 1 vs key 2, it’s more important to see accuracy plus you have the key overlay in your gameplay skin to do that anyways.
- How is it precise enough? I personally don’t think a marker as big as a cursor is precise. 

Mind you I’m not making these comments to bash, I’m just genuinely curious to know where this could go.

**Reply by Mrcubix** [reply]
> Why not make it skinnable? Is there any reason not to? Circles would be a lot easier to see considering the aim markers are already X spam, makes it weird to read.

Read again, i never said that.

> What’s the point in key 1 vs key 2, it’s more important to see accuracy plus you have the key overlay in your gameplay skin to do that anyways.

So you know which key you used to tap (ex: while trying to alt), or in case you missed an input (ex: while playing stream maps).
That functionality shouldn't deviate from rewind, if you want to see the accuracy, it's already there (300s, 100s, 50s).

> How is it precise enough? I personally don’t think a marker as big as a cursor is precise.

It's precise enough for more than just the average user, including top players.

**Reply by Sheppsu** [reply]
- I'll make the hit marker and the aim marker skinnable, including different skinnable elements for left-hit and right-hit markers; so no need to worry about their color or shape to a subjective degree (objective feedback is fine).
- I think the current hit marker is precise enough to easily tell how off you were from the circle. The entire X is big, but the intersection created by the X is quite small and precise.
- I'll add a line between the aim markers.

**Comment by Sheppsu** [comment]
**Update**
I added and improved some things. Right now, the only bug I know of is approach circles on sliders not toggling correctly when hidden is toggled on/off. Anyway, here's a [video](https://streamable.com/9afj9i) showcasing the functionality. The ending showcases that the hit markers and aim line color are skinnable. The aim markers are too, but I didn't show that in the video. If you wanna test it out for yourself, I made a [release](https://github.com/Sheppsu/osu/releases/tag/replay-anaylsis-0.1).

**Reply by OwenCMYK** [reply]
Thank you for this. I hope this feature finds it's way into the master branch at some point

**Comment by Sheppsu** [comment, marked as answer]
Drafted a PR: #27334
